vertx-mutiny-clients/vertx-mutiny-sql-client/src/main/java/io/vertx/mutiny/sqlclient/SqlClientHelper.java

```diff
@@ -0,0 +1,100 @@
+     * @return a {@link Multi} generated from operations executed inside a {@link Transaction}
+     */
+    public static <T> Multi<T> inTransactionMulti(Pool pool, Function<Transaction, Multi<T>> sourceSupplier) {
+        return usingConnectionMulti(pool, conn -> {
+            Transaction transaction = conn.begin();
+            return Multi.createBy().concatenating().streams(
+                    sourceSupplier.apply(transaction),
+                    // TODO: find a better way to transform a Uni<Void> to an empty Multi<T>
+                    transaction.commit().toMulti().transform().byTakingFirstItems(0).onItem().apply(aVoid -> (T) aVoid)).on()
+                    .failure().recoverWithMulti(throwable -> {
+                        return transaction.rollback().on().failure().recoverWithItem((Void) null)
+                                .on().item().produceMulti(v -> Multi.createFrom().failure(throwable));
+                    });
+        });
+    }
+
+    /**
+     * Generates a {@link Uni} from operations executed inside a {@link Transaction}.
+     *
+     * @param pool the {@link Pool}
+     * @param sourceSupplier a user-provided function returning a {@link Uni} generated by interacting with the given
+     *        {@link Transaction}
+     * @param <T> the type of the items emitted by the {@link Uni}
+     * @return a {@link Uni} generated from operations executed inside a {@link Transaction}
+     */
+    public static <T> Uni<T> inTransactionUni(Pool pool, Function<Transaction, Uni<T>> sourceSupplier) {
+        return usingConnectionUni(pool, conn -> {
+            Transaction transaction = conn.begin();
+            return sourceSupplier.apply(transaction)
+                    .onItem().produceUni(item -> transaction.commit().on().item().produceUni(v -> Uni.createFrom().item(item)))
+                    .on().failure().recoverWithUni(throwable -> {
+                        return transaction.rollback().on().failure().recoverWithItem((Void) null)
+                                .on().item().produceUni(v -> Uni.createFrom().failure(throwable));
+                    });
+        });
+    }
+
+    /**
+     * Generates a {@link Multi} from {@link SqlConnection} operations.
+     *
+     * @param pool the {@link Pool}
+     * @param sourceSupplier a user-provided function returning a {@link Multi} generated by interacting with the given
+     *        {@link SqlConnection}
+     * @param <T> the type of the items emitted by the {@link Multi}
+     * @return a {@link Multi} generated from {@link SqlConnection} operations
+     */
+    public static <T> Multi<T> usingConnectionMulti(Pool pool, Function<SqlConnection, Multi<T>> sourceSupplier) {
+        return pool.getConnection().onItem().produceMulti(conn -> {
+            try {
+                return sourceSupplier.apply(conn).on().termination(conn::close);
+            } catch (Throwable t) {
+                conn.close();
+                return Multi.createFrom().failure(t);
```

Comment:
Shouldn't you rollback here?

Reply:
It's not necessary. This will be invoked only if there's a runtime exception thrown during the Multi/Uni definition flow. So no actions should actually have happened.

This safety check is just here to make sure connections are released in case a dev problem was not caught before prod.